lang/funcs/core/os/exec_func.go

```diff
@@ -0,0 +1,164 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package coreos
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/purpleidea/mgmt/lang/funcs"
+	"github.com/purpleidea/mgmt/lang/interfaces"
+	"github.com/purpleidea/mgmt/lang/types"
+)
+
+func init() {
+	funcs.ModuleRegister(ModuleName, "exec", func() interfaces.Func { return &ExecFunc{} })
+}
+
+// ExecFunc runs a string as a shell command, then produces each line from
+// stdout. If the input string changes, then the commands are executed one after
+// the other and the concatenation of their outputs is produced line by line.
+//
+// Note that in the likely case in which the process emits several lines one
+// after the other, the downstream resources might not run for every line unless
+// the "Meta:realize" metaparam is set to true.
+type ExecFunc struct {
+	init *interfaces.Init
+
+	// Used to shutdown the node, killing the current process if needed.
+	// We only run one process at a time.
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// ArgGen returns the Nth arg name for this function.
+func (obj *ExecFunc) ArgGen(index int) (string, error) {
+	seq := []string{"shell_command"}
+	if l := len(seq); index >= l {
+		return "", fmt.Errorf("index %d exceeds arg length of %d", index, l)
+	}
+	return seq[index], nil
+}
+
+// Validate makes sure we've built our struct properly. It is usually unused for
+// normal functions that users can use directly.
+func (obj *ExecFunc) Validate() error {
+	return nil
+}
+
+// Info returns some static info about itself.
+func (obj *ExecFunc) Info() *interfaces.Info {
+	return &interfaces.Info{
+		Pure: false, // definitely false
+		Memo: false,
+		Sig:  types.NewType("func(shell_command str) str"),
+		Err:  obj.Validate(),
+	}
+}
+
+// Init runs some startup code for this function.
+func (obj *ExecFunc) Init(init *interfaces.Init) error {
+	obj.init = init
+	obj.ctx, obj.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *ExecFunc) Stream() error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	// Close the output chan to signal that no more values are coming.
+	defer close(obj.init.Output)
+
+	// Kill the current process if it's still running.
+	defer obj.cancel()
+
+	for {
+		select {
+		case input, more := <-obj.init.Input:
+			if !more {
+				return nil
+			}
+			shellCommand := input.Struct()["shell_command"].Str()
```

Comment:
If the input command changes (remember, input strings are a stream) we would probably want to stop exec-ing the previously running command and switch to the new one... Don't we want to kill that old one and then reset everything before running a new one here?

Reply:
Darn, what a misunderstanding!

So, this comment made me quite upset, enough that I took a break from contributing for a bit. Now that I am coming back to it, I think I understand what really happened.

I was originally planning to use the implementation you suggest in this comment, namely to kill the process when the input changes. I had told you about this plan early on, when it was still easy to change course, and I thought you had told me to pick a different course, namely to wait until the executable terminates on its own before starting the next one. I implemented what I thought you had asked, and was thus upset by the request to change it back, since I wouldn't have had to do this extra work if you let me to implement it that way to begin with.

I now understand that you were not asking me to change _my_ code so that I would send all the lines of stdout downstream, you were commenting on the fact that MCL doesn't currently send all the lines to the resources. That is, I wrote this:

```golang
// ExecFunc runs a string as a shell command, then produces each line from
// stdout. If the input string changes, the process is killed and the lines
// from the new process are produced.
//
// Note that in the likely case in which the process emits several
// lines one after the other, it is normal for some of the downstream nodes not
// to receive some of those lines. This is because MCL is not designed to
// process data streams, but rather to compute the desired state from the most
// recent values. Thus, the last line produced by the process will definitely
// be seen by the downstream nodes, and a line just before the process pauses
// is very likely to be seen as well.
```

In my mind the two paragraphs were tied: since it is only the most recent line which is guaranteed to be seen by the downstream nodes, if we know we will spawn a new process then there is no point continuing to stream the stdout of the previous process, as those lines are now known not to be the most recent value. You commented about the behaviour of MCL in the second paragraph, and I incorrectly assumed that you didn't want me to kill the old process. But I now see that you never said that and that I thus have no reason to be upset.

All this to say: I will update the code to kill the process.

Reply:
As is usual in the world, communication is always the most challenging part!

For any future things: if you disagree with whatever I'm saying (maybe I'm just phrasing it poorly and we actually agree, or perhaps I am thinking about things the wrong way) I urge you to definitely push back, as the goal is _the right way_, not necessarily _my way_. HTH and thank you for picking it back up =D